packages/datadog-instrumentations/src/mongoose.js

```diff
@@ -26,5 +26,146 @@ addHook({
+  'updateOne'
+]
+
+addHook({
+  name: 'mongoose',
+  versions: ['>=4.6.4 <5', '5', '6', '>=7'],
+  file: 'lib/model.js'
+}, Model => {
+  [...collectionMethodsWithFilter, ...collectionMethodsWithTwoFilters].forEach(methodName => {
+    const useTwoArguments = collectionMethodsWithTwoFilters.includes(methodName)
+
+    try {
+      shimmer.wrap(Model, methodName, method => {
+        return function () {
+          if (!startCh.hasSubscribers) {
+            return method.apply(this, arguments)
+          }
+
+          const asyncResource = new AsyncResource('bound-anonymous-fn')
+
+          const filters = [arguments[0]]
+          if (useTwoArguments) {
+            filters.push(arguments[1])
+          }
+
+          const finish = asyncResource.bind(function () {
+            finishCh.publish()
+          })
+
+          let callbackWrapped = false
+          const lastArgumentIndex = arguments.length - 1
+
+          if (typeof arguments[lastArgumentIndex] === 'function') {
+            // is a callback, wrap it to execute finish()
+            const originalCb = arguments[lastArgumentIndex]
+
+            arguments[lastArgumentIndex] = shimmer.wrap(originalCb, function () {
+              finish()
+
+              originalCb.apply(this, arguments)
+            })
+
+            callbackWrapped = true
+          }
+
+          return asyncResource.runInAsyncScope(() => {
+            startCh.publish({
+              filters,
+              methodName
+            })
+
+            const res = method.apply(this, arguments)
+
+            // if it is not callback, wrap exec method and its then
+            if (!callbackWrapped) {
+              const originalExec = res.exec
+
+              res.exec = shimmer.wrap(originalExec, function () {
+                const execResult = originalExec.apply(this, arguments)
+
+                // wrap them method, wrap resolve and reject methods
+                const originalThen = execResult.then
+                execResult.then = shimmer.wrap(originalThen, function () {
+                  const resolve = arguments[0]
+                  const reject = arguments[1]
+
+                  // not using shimmer here because resolve/reject could be empty
+                  arguments[0] = function () {
+                    finish()
+
+                    if (resolve) {
+                      return resolve.apply(this, arguments)
+                    }
+                  }
+
+                  arguments[1] = function () {
+                    finish()
+
+                    if (reject) {
+                      return reject.apply(this, arguments)
+                    }
+                  }
+
+                  return originalThen.apply(this, arguments)
+                })
+
+                return execResult
+              })
+            }
+            return res
+          })
+        }
+      })
+    } catch (e) {
+      // if method does not exist, do nothing
+    }
+  })
+
+  return Model
+})
```

Comment:
I'll try to separate it in few functions...

Reply:
I think that it could be uglier if I create a method like this:

```
function getWrappedThen (originalThen, finish) {
    return function wrappedThen () {
      const resolve = arguments[0]
      const reject = arguments[1]

      // not using shimmer here because resolve/reject could be empty
      arguments[0] = function wrappedResolve () {
        finish()

        if (resolve) {
          return resolve.apply(this, arguments)
        }
      }

      arguments[1] = function wrappedReject () {
        finish()

        if (reject) {
          return reject.apply(this, arguments)
        }
      }

      return originalThen.apply(this, arguments)
    }
  }
```

to be used like:
```
shimmer.wrap(execResult, 'then', function (originalThen) {
  return getWrappedThen(originalThen, finish)
})

Reply:
usually we call it `wrapX() { return function wrappedX() }`

Reply:
it's not just wrapThen btw, it's the whole 6 levels of callbacks I can see